Content.Server/GameObjects/Components/Chemistry/HyposprayComponent.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            var target = eventArgs.TargetEntity;
+            var user = eventArgs.User;
+
+            return TryDoInject(target, user);
+        }
+
+        Task IAfterInteract.AfterInteract(AfterInteractEventArgs eventArgs)
+        {
+            TryDoInject(eventArgs.Target, eventArgs.User);
+            return Task.CompletedTask;
+        }
+
+        private bool TryDoInject(IEntity? target, IEntity user)
+        {
+            if (target == null || !EligibleEntity(target))
+                return false;
+
+            var msgFormat = "You inject {0:TheName}.";
+
+            if (target == user)
+            {
+                msgFormat = "You inject yourself.";
+            }
+            else if (EligibleEntity(user) && ClumsyComponent.TryRollClumsy(user, ClumsyFailChance))
+            {
+                msgFormat = "Oops! You injected yourself!";
+                target = user;
+            }
+
+            if (_solution == null || _solution.CurrentVolume == 0)
+            {
+                user.PopupMessageCursor(Loc.GetString("It's empty!"));
+                return true;
+            }
+
+            user.PopupMessage(Loc.GetString(msgFormat, target));
+            if (target != user)
+            {
+                target.PopupMessage(Loc.GetString("You feel a tiny prick!"));
+                var meleeSys = EntitySystem.Get<MeleeWeaponSystem>();
+                var angle = new Angle(target.Transform.WorldPosition - user.Transform.WorldPosition);
+                meleeSys.SendLunge(angle, user);
+            }
+
+            EntitySystem.Get<AudioSystem>().PlayFromEntity("/Audio/Items/hypospray.ogg", user);
+
+            var targetSolution = target.GetComponent<SolutionContainerComponent>();
+
+            // Get transfer amount. May be smaller than _transferAmount if not enough room
+            var realTransferAmount = ReagentUnit.Min(TransferAmount, targetSolution.EmptyVolume);
+
+            if (realTransferAmount <= 0)
+            {
+                user.PopupMessage(user, Loc.GetString("{0:TheName} is already full!", targetSolution.Owner));
+                return true;
+            }
+
+            // Move units from attackSolution to targetSolution
+            var removedSolution = _solution.SplitSolution(realTransferAmount);
+
+            if (!targetSolution.CanAddSolution(removedSolution))
+            {
+                return true;
+            }
+
+            removedSolution.DoEntityReaction(target, ReactionMethod.Injection);
+
+            targetSolution.TryAddSolution(removedSolution);
+
+            static bool EligibleEntity(IEntity entity)
+            {
+                // TODO: Does checking for BodyComponent make sense as a "can be hypospray'd" tag?
+                // In SS13 the hypospray ONLY works on mobs, NOT beakers or anything else.
+                return entity.HasComponent<SolutionContainerComponent>() && entity.HasComponent<BodyComponent>();
```

Comment:
You can check for IBody in cases where you need to check for BodyComponent. Though no other components currently implement or register a reference for it, I use IBody where possible just in case different implementations are needed at some point.

As to what to check though, not all mobs have a body component. Xenomorphs for example don't. They do have a MobState component in common with humans and every other living thing though, which is what I usually use to distinguish mobs, however not all of them have a SolutionContainerComponent.

Reply:
I'll check BodyState then. Obviously if something doesn't have a solution container there's not a whole lot we can do with a hypospray.